pkg/startup/startup.go

```diff
@@ -246,6 +245,23 @@ func Run() {
 	}
 }
 
+func checkIPAddressSubnets(ctx context.Context, node *api.Node, cli client.Interface) bool {
```

Comment:
I think this function needs to respect the user-provided auto-detection logic. 

e.g., this line here: https://github.com/projectcalico/node/blob/99c269555fdab674f32a5087c310b1804a622033/pkg/startup/startup.go#L467

Is there a reason we can't just call this function: https://github.com/projectcalico/node/blob/99c269555fdab674f32a5087c310b1804a622033/pkg/startup/startup.go#L448?

Also looks like `ctx` and `cli` are unused, so don't need to be here.

Reply:
My original proposal called `configureIPsAndSubnets` directly. However, I noticed the following condition which looked problematic from monitoring perspective --

`} else if ipv4Env == "none" && node.Spec.BGP.IPv4Address != "" {`

If the IP address is already set which is try for monitoring case, having `ipv4Env == "none"` will just do `validateIP` and `terminate` if the IP address is not available. I have seen `ipv4Env = "none"` in a few cases (e.g. AKS) which will lead to a restart.

Should we just ignore monitoring if `ipv4Env` is set to `none`?

Reply:
Ah, hm. Yeah, I hadn't spotted that `validateIP` eventually can call terminate.

Ideally we keep the logic as close to the same as possible - I want to avoid accidental scenarios where the startup script does one thing and then this continuous code does something subtly different.

Maybe we need to do a minor refactor so that `validateIP` doesn't immediately terminate? Or we can pass a "terminationAllowed" bool through? 

Or maybe we should just pull `validateIP` out of that function altogether, and call `validateIP` after the `configureIPsAndSubnets` call exits?

We would also need to deal with this line somehow: https://github.com/projectcalico/node/blob/99c269555fdab674f32a5087c310b1804a622033/pkg/startup/startup.go#L524

Reply:
Thinking about this a bit more, we probably _also_ need to respect the collision detection logic here: https://github.com/projectcalico/node/blob/99c269555fdab674f32a5087c310b1804a622033/pkg/startup/startup.go#L187-L188

If our IP address changes, and we notice that it is the same as another node, that is a problem, and we shouldn't write the new addresses to the data store just yet. However, we should also clear our the old addresses, because we've determined that they aren't valid any more, so keeping them set would (1) be wrong and (2) potentially prevent other nodes from rightfully claiming them (e.g., if another node is assigned the address that this node used to have).